chainerrl/agents/td3.py

```diff
@@ -0,0 +1,370 @@
+
+        batch = batch_experiences(experiences, self.xp, self.phi, self.gamma)
+        self.update_q_func(batch)
+        if self.q_func1_optimizer.t % self.policy_update_delay == 0:
+            self.update_policy(batch)
+            self.sync_target_network()
+
+    def select_greedy_action(self, obs):
+        with chainer.no_backprop_mode(), chainer.using_config('train', False):
+            s = self.batch_states([obs], self.xp, self.phi)
+            action = self.policy(s).sample().array
+        return cuda.to_cpu(action)[0]
+
+    def act_and_train(self, obs, reward):
+
+        self.logger.debug('t:%s r:%s', self.t, reward)
+
+        if (self.burnin_action_func is not None
+                and self.policy_optimizer.t == 0):
+            action = self.burnin_action_func()
+        else:
+            greedy_action = self.select_greedy_action(obs)
+            action = self.explorer.select_action(self.t, lambda: greedy_action)
+        self.t += 1
+
+        if self.last_state is not None:
+            assert self.last_action is not None
+            # Add a transition to the replay buffer
+            self.replay_buffer.append(
+                state=self.last_state,
+                action=self.last_action,
+                reward=reward,
+                next_state=obs,
+                next_action=action,
+                is_state_terminal=False)
+
+        self.last_state = obs
+        self.last_action = action
+
+        self.replay_updater.update_if_necessary(self.t)
+
+        return self.last_action
+
+    def act(self, obs):
+        return self.select_greedy_action(obs)
+
+    def batch_select_greedy_action(self, batch_obs):
+        with chainer.using_config('train', False), chainer.no_backprop_mode():
+            batch_xs = self.batch_states(batch_obs, self.xp, self.phi)
+            batch_action = self.policy(batch_xs).sample().array
+        return list(cuda.to_cpu(batch_action))
+
+    def batch_act(self, batch_obs):
+        return self.batch_select_greedy_action(batch_obs)
+
+    def batch_act_and_train(self, batch_obs):
+        """Select a batch of actions for training.
+
+        Args:
+            batch_obs (Sequence of ~object): Observations.
+
+        Returns:
+            Sequence of ~object: Actions.
+        """
+
+        if (self.burnin_action_func is not None
+                and self.policy_optimizer.t == 0):
+            batch_action = [self.burnin_action_func()
+                            for _ in range(len(batch_obs))]
+        else:
+            batch_greedy_action = self.batch_select_greedy_action(batch_obs)
+            batch_action = [
+                self.explorer.select_action(
+                    self.t, lambda: batch_greedy_action[i])
+                for i in range(len(batch_greedy_action))]
+
+        self.batch_last_obs = list(batch_obs)
+        self.batch_last_action = list(batch_action)
+
+        return batch_action
+
+    def batch_observe_and_train(
```

Comment:
Can you add documentation like here:  https://github.com/chainer/chainerrl/blob/master/chainerrl/agents/ddpg.py#L409

Reply:
Since it is documented in https://github.com/chainer/chainerrl/blob/master/chainerrl/agent.py#L225, I don't think we need to repeat it here.

Reply:
Ok.